contracts/cw-ics721-bridge/src/contract.rs

```diff
@@ -0,0 +1,451 @@
+    env: Env,
+    info: MessageInfo,
+    class_id: String,
+    class_uri: Option<String>,
+    token_ids: Vec<String>,
+    token_uris: Vec<String>,
+    receiver: String,
+) -> Result<Response, ContractError> {
+    if info.sender != env.contract.address {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Optionally, instantiate a new cw721 contract if one does not
+    // yet exist.
+    let submessages = if CLASS_ID_TO_NFT_CONTRACT.has(deps.storage, class_id.clone()) {
+        let expected = CLASS_ID_TO_CLASS_URI.load(deps.storage, class_id.clone())?;
+        if expected != class_uri {
+            // classUri information for a classID shouldn't change
+            // across sends. TODO: is this the case? what are we
+            // suposed to do if not..?
+            return Err(ContractError::ClassUriClash {
+                class_id,
+                expected,
+                actual: class_uri,
+            });
+        }
+        vec![]
+    } else {
+        // Store mapping from classID to classUri. cw721 does not do
+        // this, so we need to do it to stop the infomation from
+        // getting lost.
+        CLASS_ID_TO_CLASS_URI.save(deps.storage, class_id.clone(), &class_uri)?;
+
+        let message = SubMsg::<Empty>::reply_on_success(
+            WasmMsg::Instantiate {
+                admin: None, // TODO: Any reason to set ourselves as admin?
+                code_id: CW721_ICS_CODE_ID.load(deps.storage)?,
+                msg: to_binary(
+                    &(cw721_base::msg::InstantiateMsg {
+                        // Name of the collection MUST be class_id as this is how
+                        // we create a map entry on reply.
+                        name: class_id.clone(),
+                        symbol: class_id.clone(), // TODO: What should we put here?
+                        minter: env.contract.address.to_string(),
+                    }),
+                )?,
+                funds: vec![],
+                label: format!("{} ICS721 cw721 backing contract", class_id),
+            },
+            INSTANTIATE_CW721_REPLY_ID,
+        );
+        vec![message]
+    };
+
+    // Mint the requested tokens. Submessages and their replies are
+    // always executed before regular messages [1], so we can sleep
+    // nicely knowing this won't happen until the cw721 contract has
+    // been instantiated.
+    //
+    // [1] https://github.com/CosmWasm/cosmwasm/blob/main/SEMANTICS.md#order-and-rollback
+    let mint_message = WasmMsg::Execute {
+        contract_addr: env.contract.address.into_string(),
+        msg: to_binary(&ExecuteMsg::Mint {
+            class_id,
+            token_ids,
+            token_uris,
+            receiver,
+        })?,
+        funds: vec![],
+    };
+
+    Ok(Response::default()
+        .add_attribute("method", "do_instantiate_and_mint")
+        .add_submessages(submessages)
+        .add_message(mint_message))
+}
+
+fn execute_receive_nft(
```

Comment:
Confirming that this is one of two main entry points. This is where when user does transfer nft to ics721 contract that the nft is received and then stored ( almost like a deposit nft function)
Btw we might want to use the word "deposit" somewhere here.

Reply:
Should 100% document these entry points.

Reply:
Will add some docs. This is how deposits are done. :)

Reply:
> Should 100% document these entry points.

this comment is the real MVP